File: files/CH-158.py
import pandas as pd

path = "CH-158 Filter the last transaction in mounth.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=12)
test = pd.read_excel(path, usecols="F:H", skiprows=1, nrows=4).rename(columns=lambda x: x.split('.')[0])

input['Date'] = pd.to_datetime(input['Date'])
input['month'] = input['Date'].dt.month

result = input.loc[input.groupby(['Product ID', 'month'])['Date'].idxmax()].sort_values(by='Date').reset_index(drop=True).drop(columns=['month'])

print(result.equals(test)) # True

File: files/100-199/158/CH-158.R
library(tidyverse)
library(readxl)

path = "files/CH-158 Filter the last transaction in mounth.xlsx"
input = read_excel(path, range = "B2:D14")
test  = read_excel(path, range = "F2:H6")

result = input %>% 
  group_by(`Product ID`, month = month(Date)) %>%
  filter(Date == max(Date)) %>%
  ungroup() %>%
  select(-month)

all.equal(result, test)
# [1] TRUE